src/Microsoft.VisualStudio.Sdk.TestFramework/Mocks/MockProfferAsyncService.cs

```diff
@@ -0,0 +1,72 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#if NETFRAMEWORK || WINDOWS
+
+using System.IO;
+using Microsoft;
+using Microsoft.VisualStudio.Shell.Interop;
+using ComAsyncServiceProvider = Microsoft.VisualStudio.Shell.Interop.COMAsyncServiceProvider.IAsyncServiceProvider;
+
+namespace Microsoft.VisualStudio.Sdk.TestFramework.Mocks;
+
+/// <summary>
+/// A working implementation of <see cref="IProfferAsyncService"/>.
+/// </summary>
+/// <param name="globalServiceProvider">The global service provider to which proffered services will be added.</param>
+/// <param name="logger">A means to log progress as it is reported.</param>
+public class MockProfferAsyncService(GlobalServiceProvider? globalServiceProvider, Action<string>? logger) : IProfferAsyncService
+{
+    private readonly List<Guid?> profferedCookies = new();
+
+    /// <summary>
+    /// Gets a dictionary of proffered services.
+    /// </summary>
+    public Dictionary<Guid, ComAsyncServiceProvider> ProfferedServices { get; } = new();
+
+    /// <summary>
+    /// Checks whether a service has been proffered.
+    /// </summary>
+    /// <param name="serviceType">The <see cref="Type"/> used for the service IID.</param>
+    /// <returns><see langword="true"/> if the service is proffered; otherwise <see langword="false" />.</returns>
+    public bool IsServiceProffered(Type serviceType)
+    {
+        Requires.NotNull(serviceType);
+        return this.ProfferedServices.ContainsKey(serviceType.GUID);
+    }
+
+    /// <inheritdoc/>
+    public IAsyncProgressCallback GetServiceProgressCallback()
+    {
+        return new MockAsyncProgressCallback(logger);
+    }
+
+    /// <inheritdoc/>
+    uint IProfferAsyncService.ProfferAsyncService(ref Guid rguidService, ComAsyncServiceProvider psp)
+    {
+        Guid serviceGuid = rguidService;
+        globalServiceProvider?.AddService(rguidService, async () => await psp.QueryServiceAsync(ref serviceGuid));
+        this.ProfferedServices.Add(rguidService, psp);
+        this.profferedCookies.Add(rguidService);
+        return (uint)this.profferedCookies.Count - 1;
+    }
+
+    /// <inheritdoc/>
+    void IProfferAsyncService.RevokeAsyncService(uint dwCookie)
+    {
+        Guid? guid = this.profferedCookies[(int)dwCookie];
```

Comment:
Technically the shell proffer service silently ignores revokes on cookies that don't exist in our map

Reply:
That sounds like it lets bugs go silently. Interesting. If someone has lost track of their cookies, they might also unproffer someone else's service by accident.
I think I prefer throwing. At least in testing then, a bug in their product will be caught, which should accrue value to their shipping software.